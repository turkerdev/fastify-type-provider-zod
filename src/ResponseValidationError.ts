import type { z, ZodIssue } from 'zod';

export type ResponseValidationErrorDetails = {
  error: ZodIssue[];
  method: string;
  url: string;
};

export class ResponseValidationError extends Error {
  public details: ResponseValidationErrorDetails;

  constructor(
    validationResult: z.SafeParseReturnType<unknown, unknown>,
    method: string,
    url: string,
  ) {
    super("Response doesn't match the schema");
    this.name = 'ResponseValidationError';
    this.details = {
      error: validationResult.error?.issues ?? [],
      method,
      url,
    };
  }
}
