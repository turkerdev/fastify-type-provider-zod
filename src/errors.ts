import createError from '@fastify/error';
import type { FastifySchemaValidationError } from 'fastify/types/schema';
import type { ZodError } from 'zod';

export class ResponseSerializationError extends createError<[{ cause: ZodError }]>(
  'FST_ERR_RESPONSE_SERIALIZATION',
  "Response doesn't match the schema",
  500,
) {
  constructor(
    public cause: ZodError,
    public method: string,
    public url: string,
  ) {
    super({ cause });
  }
}

export const InvalidSchemaError = createError<[string]>(
  'FST_ERR_INVALID_SCHEMA',
  'Invalid schema passed: %s',
  500,
);

export const createValidationError = (
  error: ZodError,
  method: string,
  url: string,
): FastifySchemaValidationError[] =>
  error.errors.map((issue) => ({
    keyword: issue.code,
    instancePath: `/${issue.path.join('/')}`,
    schemaPath: `#/${issue.path.join('/')}/${issue.code}`,
    params: {
      issue,
      zodError: error,
      method,
      url,
    },
    message: error.message,
  }));
