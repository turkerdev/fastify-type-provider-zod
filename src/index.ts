import type {
  FastifyPluginAsync,
  FastifyPluginCallback,
  FastifyPluginOptions,
  FastifySchema,
  FastifySchemaCompiler,
  FastifyTypeProvider,
  RawServerBase,
  RawServerDefault,
} from 'fastify';
import type { FastifySerializerCompiler } from 'fastify/types/schema';
import type { ZodAny, ZodTypeAny, z } from 'zod';
import { zodToJsonSchema, defaultOptions } from 'zod-to-json-schema';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type FreeformRecord = Record<string, any>;

const defaultSkipList = [
  '/documentation/',
  '/documentation/initOAuth',
  '/documentation/json',
  '/documentation/uiConfig',
  '/documentation/yaml',
  '/documentation/*',
  '/documentation/static/*',
];

export interface ZodTypeProvider extends FastifyTypeProvider {
  output: this['input'] extends ZodTypeAny ? z.infer<this['input']> : never;
}

interface Schema extends FastifySchema {
  hide?: boolean;
}

type createJsonSchemaTransformOptions = Partial<typeof defaultOptions> & {
  skipList: readonly string[];
}

const zodToJsonSchemaOptions = {
  target: 'openApi3',
  $refStrategy: 'none',
} as const;

export const createJsonSchemaTransform = ({ skipList, ..._options }: createJsonSchemaTransformOptions) => {
  return ({ schema, url }: { schema: Schema; url: string }) => {
    if (!schema) {
      return {
        schema,
        url,
      };
    }

    const options = { ...zodToJsonSchemaOptions, ..._options };

    const { response, headers, querystring, body, params, hide, ...rest } = schema;

    const transformed: FreeformRecord = {};

    if (skipList.includes(url) || hide) {
      transformed.hide = true;
      return { schema: transformed, url };
    }

    const zodSchemas: FreeformRecord = { headers, querystring, body, params };

    for (const prop in zodSchemas) {
      const zodSchema = zodSchemas[prop];
      if (zodSchema) {
        transformed[prop] = zodToJsonSchema(zodSchema, options);
      }
    }

    if (response) {
      transformed.response = {};

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      for (const prop in response as any) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const schema = resolveSchema((response as any)[prop]);

        const transformedResponse = zodToJsonSchema(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          schema as any,
          options,
        );
        transformed.response[prop] = transformedResponse;
      }
    }

    for (const prop in rest) {
      const meta = rest[prop as keyof typeof rest];
      if (meta) {
        transformed[prop] = meta;
      }
    }

    return { schema: transformed, url };
  };
};

export const jsonSchemaTransform = createJsonSchemaTransform({
  skipList: defaultSkipList,
});

export const validatorCompiler: FastifySchemaCompiler<ZodAny> =
  ({ schema }) =>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (data): any => {
    try {
      return { value: schema.parse(data) };
    } catch (error) {
      return { error };
    }
  };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function hasOwnProperty<T, K extends PropertyKey>(obj: T, prop: K): obj is T & Record<K, any> {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

function resolveSchema(maybeSchema: ZodAny | { properties: ZodAny }): Pick<ZodAny, 'safeParse'> {
  if (hasOwnProperty(maybeSchema, 'safeParse')) {
    return maybeSchema;
  }
  if (hasOwnProperty(maybeSchema, 'properties')) {
    return maybeSchema.properties;
  }
  throw new Error(`Invalid schema passed: ${JSON.stringify(maybeSchema)}`);
}

export class ResponseValidationError extends Error {
  public details: FreeformRecord;

  constructor(validationResult: FreeformRecord) {
    super("Response doesn't match the schema");
    this.name = 'ResponseValidationError';
    this.details = validationResult.error;
  }
}

export const serializerCompiler: FastifySerializerCompiler<ZodAny | { properties: ZodAny }> =
  ({ schema: maybeSchema }) =>
  (data) => {
    const schema: Pick<ZodAny, 'safeParse'> = resolveSchema(maybeSchema);

    const result = schema.safeParse(data);
    if (result.success) {
      return JSON.stringify(result.data);
    }

    throw new ResponseValidationError(result);
  };

/**
 * FastifyPluginCallbackZod with Zod automatic type inference
 *
 * @example
 * ```typescript
 * import { FastifyPluginCallbackZod } fromg "fastify-type-provider-zod"
 *
 * const plugin: FastifyPluginCallbackZod = (fastify, options, done) => {
 *   done()
 * }
 * ```
 */
export type FastifyPluginCallbackZod<
  Options extends FastifyPluginOptions = Record<never, never>,
  Server extends RawServerBase = RawServerDefault,
> = FastifyPluginCallback<Options, Server, ZodTypeProvider>;

/**
 * FastifyPluginAsyncZod with Zod automatic type inference
 *
 * @example
 * ```typescript
 * import { FastifyPluginAsyncZod } fromg "fastify-type-provider-zod"
 *
 * const plugin: FastifyPluginAsyncZod = async (fastify, options) => {
 * }
 * ```
 */
export type FastifyPluginAsyncZod<
  Options extends FastifyPluginOptions = Record<never, never>,
  Server extends RawServerBase = RawServerDefault,
> = FastifyPluginAsync<Options, Server, ZodTypeProvider>;
