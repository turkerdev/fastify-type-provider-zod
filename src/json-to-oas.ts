import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types'
import type { JSONSchema } from 'zod/v4/core'

type OASVersion = '3.0' | '3.1'

export const getOASVersion = (documentObject: {
  openapiObject: Partial<OpenAPIV3.Document | OpenAPIV3_1.Document>
}): OASVersion => {
  const openapiVersion = documentObject.openapiObject.openapi || '3.0.3'

  if (openapiVersion.startsWith('3.1')) {
    return '3.1'
  }

  if (openapiVersion.startsWith('3.0')) {
    return '3.0'
  }

  throw new Error('Unsupported OpenAPI document object')
}

export const jsonSchemaToOAS_3_0 = (jsonSchema: JSONSchema.BaseSchema): OpenAPIV3.SchemaObject => {
  const clone: any = { ...jsonSchema }

  if (clone.type === 'null') {
    clone.nullable = true
    delete clone.type
    clone.enum = [null]
  }

  if (Array.isArray(clone.prefixItems)) {
    const tuple = clone.prefixItems as JSONSchema.BaseSchema[]

    clone.minItems ??= tuple.length
    clone.maxItems ??= tuple.length

    clone.items = {
      oneOf: tuple.map(jsonSchemaToOAS_3_0),
    }

    delete clone.prefixItems
  }

  if ('const' in clone && clone.const !== undefined) {
    clone.enum = [clone.const]
    delete clone.const
  }

  if (typeof clone.exclusiveMinimum === 'number') {
    clone.minimum = clone.exclusiveMinimum
    clone.exclusiveMinimum = true
  }
  if (typeof clone.exclusiveMaximum === 'number') {
    clone.maximum = clone.exclusiveMaximum
    clone.exclusiveMaximum = true
  }

  for (const key of [
    '$schema',
    '$id',
    'unevaluatedProperties',
    'dependentSchemas',
    'patternProperties',
    'propertyNames',
    'contentEncoding',
    'contentMediaType',
  ]) {
    delete clone[key]
  }

  const recursive = (v: any): any =>
    Array.isArray(v) ? v.map(jsonSchemaToOAS_3_0) : jsonSchemaToOAS_3_0(v)

  if (clone.properties) {
    for (const [k, v] of Object.entries(clone.properties)) {
      clone.properties![k] = jsonSchemaToOAS_3_0(v as any)
    }
  }

  if (clone.items && !Array.isArray(clone.items)) {
    clone.items = recursive(clone.items)
  }

  for (const key of ['allOf', 'anyOf', 'oneOf', 'not', 'then', 'else', 'if', 'contains']) {
    if (clone[key]) {
      clone[key] = recursive(clone[key])
    }
  }

  return clone as OpenAPIV3.SchemaObject
}

const jsonSchemaToOAS_3_1 = (jsonSchema: JSONSchema.BaseSchema): OpenAPIV3_1.SchemaObject => {
  return jsonSchema as OpenAPIV3_1.SchemaObject
}

export const jsonSchemaToOAS = (
  jsonSchema: JSONSchema.BaseSchema,
  oasVersion: OASVersion,
): OpenAPIV3.SchemaObject | OpenAPIV3_1.SchemaObject => {
  switch (oasVersion) {
    case '3.0':
      return jsonSchemaToOAS_3_0(jsonSchema)
    case '3.1':
      return jsonSchemaToOAS_3_1(jsonSchema)
    default:
      throw new Error(`Unsupported OpenAPI version: ${oasVersion}`)
  }
}
