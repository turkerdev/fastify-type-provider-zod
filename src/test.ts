import z, { toJSONSchema } from 'zod'

console.log(toJSONSchema(z.union([z.string(), z.number()]).array()))
// {
//   '$schema': 'https://json-schema.org/draft/2020-12/schema',
//   type: 'array',
//   items: { anyOf: [ [Object], [Object] ] }
// }

console.log(toJSONSchema(z.tuple([z.string(), z.number()])))
// {
//   '$schema': 'https://json-schema.org/draft/2020-12/schema',
//   type: 'array',
//   prefixItems: [ { type: 'string' }, { type: 'number' } ]
// }
