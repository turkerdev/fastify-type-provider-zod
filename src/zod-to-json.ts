import type { z } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';

const zodToJsonSchemaOptions = {
  target: 'openApi3',
  $refStrategy: 'none',
} as const;

export const convertZodToJsonSchema = (zodSchema: z.ZodTypeAny) => {
  return zodToJsonSchema(zodSchema, zodToJsonSchemaOptions);
};
