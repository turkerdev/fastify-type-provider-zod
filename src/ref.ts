import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';
import type { z } from 'zod';

import { convertZodToJsonSchema } from './zod-to-json';

const createComponentMap = (
  schemas: Record<string, OpenAPIV3.SchemaObject | OpenAPIV3_1.SchemaObject>,
) => {
  const map = new Map<string, string>();

  Object.entries(schemas).forEach(([key, value]) => map.set(JSON.stringify(value), key));

  return map;
};

const createComponentReplacer = (componentMapVK: Map<string, string>, schemasObject: object) =>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function componentReplacer(this: any, key: string, value: any) {
    if (typeof value !== 'object') return value;

    // Check if the parent is the schemas object, if so, return the value as is. This is where the schemas are defined.
    if (this === schemasObject) return value;

    const stringifiedValue = JSON.stringify(value);
    if (componentMapVK.has(stringifiedValue))
      return { $ref: `#/components/schemas/${componentMapVK.get(stringifiedValue)}` };

    if (value.nullable === true) {
      const nonNullableValue = { ...value };
      delete nonNullableValue.nullable;
      const stringifiedNonNullableValue = JSON.stringify(nonNullableValue);
      if (componentMapVK.has(stringifiedNonNullableValue))
        return {
          anyOf: [
            { $ref: `#/components/schemas/${componentMapVK.get(stringifiedNonNullableValue)}` },
          ],
          nullable: true,
        };
    }

    return value;
  };

export const resolveRefs = (
  openapiObject: Partial<OpenAPIV3.Document | OpenAPIV3_1.Document>,
  zodSchemas: Record<string, z.ZodTypeAny>,
) => {
  const schemas: Record<string, OpenAPIV3.SchemaObject | OpenAPIV3_1.SchemaObject> = {};
  for (const key in zodSchemas) {
    schemas[key] = convertZodToJsonSchema(zodSchemas[key]);
  }

  const document = {
    ...openapiObject,
    components: {
      ...openapiObject.components,
      schemas: {
        ...openapiObject.components?.schemas,
        ...schemas,
      },
    },
  };

  const componentMapVK = createComponentMap(schemas);
  const componentReplacer = createComponentReplacer(componentMapVK, document.components.schemas);

  // Using the componentReplacer function we deep check if the document has any schemas that are the same as the zod schemas provided
  // When a match is found replace them with a $ref.
  return JSON.parse(JSON.stringify(document, componentReplacer));
};
