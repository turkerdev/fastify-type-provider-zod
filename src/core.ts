import type {
  FastifyPluginAsync,
  FastifyPluginCallback,
  FastifyPluginOptions,
  FastifySchema,
  FastifySchemaCompiler,
  FastifyTypeProvider,
  RawServerBase,
  RawServerDefault,
} from 'fastify';
import type { FastifySerializerCompiler } from 'fastify/types/schema';
import type { OpenAPIV2, OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';
import type { z } from 'zod';

import { createValidationError, InvalidSchemaError, ResponseSerializationError } from './errors';
import { resolveRefs } from './ref';
import { convertZodToJsonSchema } from './zod-to-json';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type FreeformRecord = Record<string, any>;

const defaultSkipList = [
  '/documentation/',
  '/documentation/initOAuth',
  '/documentation/json',
  '/documentation/uiConfig',
  '/documentation/yaml',
  '/documentation/*',
  '/documentation/static/*',
];

export interface ZodTypeProvider extends FastifyTypeProvider {
  validator: this['schema'] extends z.ZodTypeAny ? z.output<this['schema']> : unknown;
  serializer: this['schema'] extends z.ZodTypeAny ? z.input<this['schema']> : unknown;
}

interface Schema extends FastifySchema {
  hide?: boolean;
}

export const createJsonSchemaTransform = ({ skipList }: { skipList: readonly string[] }) => {
  return ({ schema, url }: { schema: Schema; url: string }) => {
    if (!schema) {
      return {
        schema,
        url,
      };
    }

    const { response, headers, querystring, body, params, hide, ...rest } = schema;

    const transformed: FreeformRecord = {};

    if (skipList.includes(url) || hide) {
      transformed.hide = true;
      return { schema: transformed, url };
    }

    const zodSchemas: FreeformRecord = { headers, querystring, body, params };

    for (const prop in zodSchemas) {
      const zodSchema = zodSchemas[prop];
      if (zodSchema) {
        transformed[prop] = convertZodToJsonSchema(zodSchema);
      }
    }

    if (response) {
      transformed.response = {};

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      for (const prop in response as any) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const schema = resolveSchema((response as any)[prop]);

        const transformedResponse = convertZodToJsonSchema(schema);
        transformed.response[prop] = transformedResponse;
      }
    }

    for (const prop in rest) {
      const meta = rest[prop as keyof typeof rest];
      if (meta) {
        transformed[prop] = meta;
      }
    }

    return { schema: transformed, url };
  };
};

export const jsonSchemaTransform = createJsonSchemaTransform({
  skipList: defaultSkipList,
});

export const createJsonSchemaTransformObject =
  ({ schemas }: { schemas: Record<string, z.ZodTypeAny> }) =>
  (
    input:
      | { swaggerObject: Partial<OpenAPIV2.Document> }
      | { openapiObject: Partial<OpenAPIV3.Document | OpenAPIV3_1.Document> },
  ) => {
    if ('swaggerObject' in input) {
      console.warn('This package currently does not support component references for Swagger 2.0');
      return input.swaggerObject;
    }

    return resolveRefs(input.openapiObject, schemas);
  };

export const validatorCompiler: FastifySchemaCompiler<z.ZodTypeAny> =
  ({ schema, method, url }) =>
  (data) => {
    const result = schema.safeParse(data);
    if (result.error) {
      return { error: createValidationError(result.error, method, url) as unknown as Error };
    }

    return { value: result.data };
  };

function resolveSchema(maybeSchema: z.ZodTypeAny | { properties: z.ZodTypeAny }): z.ZodTypeAny {
  if ('safeParse' in maybeSchema) {
    return maybeSchema;
  }
  if ('properties' in maybeSchema) {
    return maybeSchema.properties;
  }
  throw new InvalidSchemaError(JSON.stringify(maybeSchema));
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ReplacerFunction = (this: any, key: string, value: any) => any;

export type ZodSerializerCompilerOptions = {
  replacer?: ReplacerFunction;
};

export const createSerializerCompiler =
  (
    options?: ZodSerializerCompilerOptions,
  ): FastifySerializerCompiler<z.ZodTypeAny | { properties: z.ZodTypeAny }> =>
  ({ schema: maybeSchema, method, url }) =>
  (data) => {
    const schema = resolveSchema(maybeSchema);

    const result = schema.safeParse(data);
    if (result.error) {
      throw new ResponseSerializationError(result.error, method, url);
    }

    return JSON.stringify(result.data, options?.replacer);
  };

export const serializerCompiler = createSerializerCompiler({});

/**
 * FastifyPluginCallbackZod with Zod automatic type inference
 *
 * @example
 * ```typescript
 * import { FastifyPluginCallbackZod } from "fastify-type-provider-zod"
 *
 * const plugin: FastifyPluginCallbackZod = (fastify, options, done) => {
 *   done()
 * }
 * ```
 */
export type FastifyPluginCallbackZod<
  Options extends FastifyPluginOptions = Record<never, never>,
  Server extends RawServerBase = RawServerDefault,
> = FastifyPluginCallback<Options, Server, ZodTypeProvider>;

/**
 * FastifyPluginAsyncZod with Zod automatic type inference
 *
 * @example
 * ```typescript
 * import { FastifyPluginAsyncZod } from "fastify-type-provider-zod"
 *
 * const plugin: FastifyPluginAsyncZod = async (fastify, options) => {
 * }
 * ```
 */
export type FastifyPluginAsyncZod<
  Options extends FastifyPluginOptions = Record<never, never>,
  Server extends RawServerBase = RawServerDefault,
> = FastifyPluginAsync<Options, Server, ZodTypeProvider>;
