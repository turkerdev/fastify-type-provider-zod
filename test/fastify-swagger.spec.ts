import fastifySwagger from '@fastify/swagger';
import fastifySwaggerUI from '@fastify/swagger-ui';
import Fastify from 'fastify';
import * as validator from 'oas-validator';
import { z } from 'zod';

import type { ZodTypeProvider } from '../src';
import { jsonSchemaTransform, serializerCompiler, validatorCompiler } from '../src';

describe('transformer', () => {
  it('generates types for fastify-swagger correctly', async () => {
    const app = Fastify();
    app.setValidatorCompiler(validatorCompiler);
    app.setSerializerCompiler(serializerCompiler);

    app.register(fastifySwagger, {
      openapi: {
        info: {
          title: 'SampleApi',
          description: 'Sample backend service',
          version: '1.0.0',
        },
        servers: [],
      },
      transform: jsonSchemaTransform,
    });

    app.register(fastifySwaggerUI, {
      routePrefix: '/documentation',
    });

    const LOGIN_SCHEMA = z.object({
      username: z.string().max(32).describe('someDescription'),
      seed: z.number().positive(),
      password: z.string().max(32),
    });

    const UNAUTHORIZED_SCHEMA = z.object({
      required_role: z.literal('admin'),
    });

    app.after(() => {
      app
        .withTypeProvider<ZodTypeProvider>()
        .route({
          method: 'POST',
          url: '/login',
          schema: {
            description: 'login route',
            summary: 'login your account',
            consumes: ['application/json'],
            deprecated: false,
            hide: false,
            tags: ['auth'],
            externalDocs: { url: 'https://google.com', description: 'check google' },
            body: LOGIN_SCHEMA,
            response: {
              200: z.string(),
              401: UNAUTHORIZED_SCHEMA,
            },
          },
          handler: (req, res) => {
            res.send('ok');
          },
        })
        .route({
          method: 'POST',
          url: '/no-schema',
          schema: undefined,
          handler: (req, res) => {
            res.send('ok');
          },
        })
        .route({
          method: 'DELETE',
          url: '/delete',
          schema: {
            description: 'delete route',
            response: {
              204: z.undefined().describe('Empty response'),
            },
          },
          handler: (req, res) => {
            res.status(204).send();
          },
        });
    });

    await app.ready();

    const openApiSpecResponse = await app.inject().get('/documentation/json');
    const openApiSpec = JSON.parse(openApiSpecResponse.body);

    expect(openApiSpec).toMatchSnapshot();
    await validator.validate(openApiSpec, {});
  });
});
