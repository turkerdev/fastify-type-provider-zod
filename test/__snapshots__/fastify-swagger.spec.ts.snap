// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformer > generates types for fastify-swagger correctly 1`] = `
{
  "components": {
    "schemas": {},
  },
  "info": {
    "description": "Sample backend service",
    "title": "SampleApi",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": {
    "/delete": {
      "delete": {
        "description": "delete route",
        "responses": {
          "204": {
            "content": {
              "application/json": {
                "schema": {
                  "description": "Empty response",
                  "not": {},
                },
              },
            },
            "description": "Empty response",
          },
        },
      },
    },
    "/login": {
      "post": {
        "description": "login route",
        "externalDocs": {
          "description": "check google",
          "url": "https://google.com",
        },
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "password": {
                    "maxLength": 32,
                    "type": "string",
                  },
                  "seed": {
                    "exclusiveMinimum": true,
                    "minimum": 0,
                    "type": "number",
                  },
                  "username": {
                    "description": "someDescription",
                    "maxLength": 32,
                    "type": "string",
                  },
                },
                "required": [
                  "username",
                  "seed",
                  "password",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "type": "string",
                },
              },
            },
            "description": "Default Response",
          },
          "401": {
            "content": {
              "application/json": {
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "required_role": {
                      "enum": [
                        "admin",
                      ],
                      "type": "string",
                    },
                  },
                  "required": [
                    "required_role",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default Response",
          },
        },
        "summary": "login your account",
        "tags": [
          "auth",
        ],
      },
    },
    "/no-schema": {
      "post": {
        "responses": {
          "200": {
            "description": "Default Response",
          },
        },
      },
    },
  },
  "servers": [],
}
`;

exports[`transformer > should generate ref correctly 1`] = `
{
  "components": {
    "schemas": {
      "Token": {
        "maxLength": 12,
        "minLength": 12,
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Sample backend service",
    "title": "SampleApi",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": {
    "/login": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "access_token": {
                    "$ref": "#/components/schemas/Token",
                  },
                  "refresh_token": {
                    "$ref": "#/components/schemas/Token",
                  },
                },
                "required": [
                  "access_token",
                  "refresh_token",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "description": "Default Response",
          },
        },
      },
    },
  },
  "servers": [],
}
`;

exports[`transformer > should not generate ref 1`] = `
{
  "components": {
    "schemas": {},
  },
  "info": {
    "description": "Sample backend service",
    "title": "SampleApi",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": {
    "/login": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "additionalProperties": false,
                "properties": {
                  "access_token": {
                    "maxLength": 12,
                    "minLength": 12,
                    "type": "string",
                  },
                  "refresh_token": {
                    "maxLength": 12,
                    "minLength": 12,
                    "type": "string",
                  },
                },
                "required": [
                  "access_token",
                  "refresh_token",
                ],
                "type": "object",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "description": "Default Response",
          },
        },
      },
    },
  },
  "servers": [],
}
`;
